apiBase/APIRequestUtils.es6.js

```diff
@@ -99,72 +63,30 @@ export const validateData = (data, method, apiName, validator) => {
 export const save = (apiOptions, method, path, data, parseBody, parseMeta) => {
   if (!(apiOptions && method && path && data)) { throw new NoModelError(); }
 
-  return new Promise((resolve, reject) => {
-    rawSend(apiOptions, method, path, data, 'form', (err, res, req) => {
-      handle(apiOptions, resolve, reject, err, res, req, data, method, parseBody, parseMeta);
-    });
-  });
-};
-
-export const runQuery = (apiOptions, method, path, query, rawQuery, parseBody, parseMeta) => {
-  if (!(apiOptions && method && path && query && rawQuery)) { throw new NoModelError(); }
-
-  if (method === 'get') {
-    query.raw_json = 1;
-  }
-
-  return new Promise((resolve, reject) => {
-    rawSend(apiOptions, method, path, query, 'query', (err, res, req) => {
-      handle(apiOptions, resolve, reject, err, res, req, rawQuery, method, parseBody, parseMeta);
-    });
-  });
+  return rawSend(apiOptions, method, path, data, 'form')
+    .then((res, req) => makeApiResponse(res, req, method, query, parseBody, parseMeta))
+    .catch((err, req) => console.trace(err));
 };
 
-const normalizeRequest = (res, req) => {
-  if (res && !req) {
-    return res.request || res.req;
+export const runQuery = (apiOptions, method, path, query, parseBody, parseMeta) => {
+  if (!(apiOptions && method && path && query)) {
+    throw new NoModelError();     // TODO: does this error make sense?
   }
 
-  return req;
-};
-
-const handleRequestIfFailed = (apiOptions, err, res, req, reject) => {
-  if (!(err || (res && !res.ok))) { return; }
-
-  getEmitter(apiOptions).emit(Events.error, err, req);
-  const errorHandler = apiOptions.defaultErrorHandler || reject;
-  errorHandler(err || 500);
-  return true;
-};
-
-const handle = (apiOptions, resolve, reject, err, res, req, query, method,
-    parseBody, parseMeta) => {
-
-  req = normalizeRequest(res, req);
-
-  if (handleRequestIfFailed(apiOptions, err, res, req, reject)) {
-    return;
+  if (method.toLowerCase() === 'get') {
+    query.raw_json = 1;
   }
 
-  getEmitter(apiOptions).emit(Events.response, req, res);
-
-  const apiResponse = tryParseResponse(reject, res, req, method, query, parseBody, parseMeta);
-
-  getEmitter(apiOptions).emit(Events.result, apiResponse);
-  resolve(apiResponse);
+  return rawSend(apiOptions, method, path, query, 'query')
+    .then((res, req) => makeApiResponse(res, req, method, query, parseBody, parseMeta))
+    .catch((err, req) => console.trace(err));
```

Comment:
So I tacked that on there for feature parity but I'm glad you caught that. Ideally, the calling code applies the catch so it can be handled more precisely there. The problem with that, as I'm sure you know, is that Promises will silently squash exceptions if you forget to add a `catch`. I'm not sure how/if we've remedied that somehow.